### ISSUE TITLE
XCTAssertNotNil cannot handle optionals

### ISSUE THREAD

**Issue by flashspys** [SwifterSwift/SwifterSwift#188]
In your tests you are checking at various locations optionals not to be nil by passing them to the `XCTAssertNotNil` function. Unfortunally the function is unable to process Swift optionals and will threat even nil-optionals as not-nil. For this reason #187 is crashing and not failing, because first the optional is checked and then force-unwrapped. Another option to check optionals to be nil should be considered, [for example as mentioned in this blogpost](https://www.raizlabs.com/dev/2017/02/xctest-optional-unwrapping/).

**Comment by flashspys** [comment]
Another option would be to add SwifterSwift's very first XCTest extension 😆 

`XCTest.assertOptionalNotNil`

**Comment by SD10** [comment]
@flashspys Thanks for pointing this out. I think we could also set XCTest to stop testing after failing once. The downside to that is we don't know of any failed tests after this until the next run, but I think it could be a fair compromise.

**Comment by flashspys** [comment]
This will not solve the situation in that a single test is crashing ( != failing) because some force-unwrap fails.

**Comment by SD10** [comment]
@flashspys let me make sure I understand correctly, are you saying this test won't fail:
```swift
let nilValue: Int? = nil
XCTAssertNotNil(nilValue)
```

**Comment by flashspys** [comment]
No, the behaviour is more complicated. I think the following picture will help to understand:

In this case, `XCTAssertNotNil` will fail as expected.
<img width="461" alt="screen shot 2017-07-08 at 23 38 57" src="https://user-images.githubusercontent.com/1114404/27989346-da3334ee-6436-11e7-8bcb-6b6e5f528431.png">


But in this case `XCTAssertNotNil` will not detect the nilness of `string`. The test will crash at the force-unwrap
<img width="445" alt="screen shot 2017-07-08 at 23 39 51" src="https://user-images.githubusercontent.com/1114404/27989345-da2ef1ae-6436-11e7-83da-44eefcda0a26.png">

Whyever...

**Comment by SD10** [comment]
@flashspys Ok, I think we're on the same page. That is why I was suggesting this:

```Swift
  override func setUp() {
    continueAfterFailure = false
  }
```
If `XCTAssertNotNil` fails the proceeding code won't be executed.

**Comment by flashspys** [comment]
I think it is more an UI Bug than I thought because your idea is right. It will fix it and even in the second example it will stop executing after `XCTAssertNotNil` when you add `continueAfterFailure = false`
<img width="442" alt="screen shot 2017-07-09 at 00 00 29" src="https://user-images.githubusercontent.com/1114404/27989463-a8199428-6439-11e7-8cdf-a1b177f7ac3b.png">
My guess is that Xcode is just not displaying these small red x in case of a runtime error.

And now I understand your idea from your first answer. I'm happy to announce that we don't even need to make a compromise. [You can stop a single test after fail without stop the other tests](https://stackoverflow.com/a/27016786/1857527).

**Comment by SD10** [comment]
@flashspys Yeah the method I showed above would work. What's happening is there is no test report being shown due to the crash. `XCTAssertNotNil` is still detecting the nilness of the `String`.

![screen shot 2017-07-08 at 5 03 25 pm](https://user-images.githubusercontent.com/7445580/27989494-64156246-63ff-11e7-81f7-0544c0f5959b.png)

Feel free to open a Pull Request making these changes to the tests where we use `XCTAssertNotNil`. In general, this is a solid example as to why we should stop using force unwrapping even in tests.

**Comment by omaralbeik** [comment]
Fixed in v3.1.0
